README.md
TimeShift.js
============

[![npm](https://img.shields.io/npm/v/timeshift-js.svg)](https://www.npmjs.com/package/timeshift-js)

TimeShift.js allows mocking / overriding JavaScript's Date object so that you can set the current time and timezone.  It is meant for creating repeatable tests that utilize the current time or date.

Usage
-----

```javascript
new Date().toString();                      // Original Date object
"Fri Aug 09 2013 23:37:42 GMT+0300 (EEST)"

Date = TimeShift.Date;                      // Overwrite Date object
new Date().toString();
"Fri Aug 09 2013 23:37:43 GMT+0300"

TimeShift.setTimezoneOffset(-60);           // Set timezone to GMT+0100 (note the sign)
new Date().toString();
"Fri Aug 09 2013 21:37:44 GMT+0100"

TimeShift.setTime(1328230923000);           // Set the time to 2012-02-03 01:02:03 GMT
new Date().toString();
"Fri Feb 03 2012 02:02:03 GMT+0100"

TimeShift.setTimezoneOffset(0);             // Set timezone to GMT
new Date().toString();
"Fri Feb 03 2012 01:02:03 GMT"

TimeShift.getTime();                        // Get overridden values
1328230923000
TimeShift.getTimezoneOffset();
0

var t = 1577836800000;
TimeShift.setTime(() => {                   // Set the time using a callback that's evaluated each time a new Date is created
    t += 1000;
    return t;
});
new Date().toString();
"Wed Jan 01 2020 00:00:01 GMT"
new Date().toString();
"Wed Jan 01 2020 00:00:02 GMT"

TimeShift.setTime(undefined);               // Reset to current time
new Date().toString();
"Fri Aug 09 2013 20:37:45 GMT"

new Date().desc();                          // Helper method
"utc=Fri, 09 Aug 2013 20:37:46 GMT   local=Fri, 09 Aug 2013 20:37:46 GMT   offset=0"

new TimeShift.OriginalDate().toString();    // Use original Date object
"Fri Aug 09 2013 23:37:47 GMT+0300 (EEST)"
```

Time zones
----------

TimeShift.js always utilizes its internal time zone offset when converting between local time and UTC.  The offset factor is fixed, and it does not take into account DST changes.  Effectively it emulates a time zone with no DST.

```javascript
new Date(1370034000000).toString();         // Original Date object uses variable offset
"Sat Jun 01 2013 00:00:00 GMT+0300 (EEST)"
new Date(1356991200000).toString();
"Tue Jan 01 2013 00:00:00 GMT+0200 (EET)"

Date = TimeShift.Date;                      // TimeShift.js uses fixed offset
new Date(1370034000000).toString();
"Sat Jun 01 2013 00:00:00 GMT+0300"
new Date(1356991200000).toString();
"Tue Jan 01 2013 01:00:00 GMT+0300"
```

The default time zone offset is the current local time zone offset.  Note that this can change depending on local DST.  Setting the time zone offset affects also previously created Date instances.

The time zone offset has the same sign as [Date.getTimezoneOffset](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTimezoneOffset).  For example, -120 is GMT+0200 and +120 is GMT-0200.

Mocking Passage of Time
-----------------------
Once the time is set, the same time will be returned every time you create a new Date object by default. It will not progress with the passage of time. Most of the time this is the desired behavior, but sometimes this can cause issues with things like canvas animations. If you want to maintain the normal passage of time starting with the date you set you can do so as follows:

```javascript
Date = TimeShift.Date;                                        // Override the Date object as usual
var originalDate = new TimeShift.OriginalDate().getTime();    // Get the actual date before setting
var mockTimestamp = 1577836800000;

// Pass a callback to setTime that adds the change in time since the date was mocked to the the mocked time.
TimeShift.setTime(() => {
    var dateNow = new TimeShift.OriginalDate().getTime();
    return mockTimestamp + dateNow - originalDate;
});

new Date().toString()
"Wed Jan 01 2020 00:00:01 GMT"
new Date().toString()
"Wed Jan 01 2020 00:00:02 GMT"
new Date().toString()
"Wed Jan 01 2020 00:00:03 GMT"
```

Caveats
-------

The mock implementation of Date is not perfect.

* Many string-generation methods are incomplete and return something indicative, but not fully correct.  In particular `toDateString`, `toLocaleDateString`, `toLocaleString`, `toLocaleTimeString`, `toTimeString` produce somewhat incorrect results.

* The `toString` method does not contain any time zone name.

* The `parse` method delegates directly to the original method and may not handle time zones correctly.

* DST changes cannot be emulated.  The time zone offset it always fixed.

* If a library or other code holds an original Date object or a reference to the Date prototype, things may break (e.g. error messages like "this is not a Date object").  In this case you should overwrite the Date object before loading the library.

If you'd like to fix some of these issues, please fork the repository, implement the desired functionality, add unit tests to `tests.js` and send a pull request.

License
-------

TimeShift.js is Copyright 2013 Mobile Wellness Solutions MWS Ltd and Sampo Juustila.

It is released under the MIT license.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.2.0] - 2022-06-01

### Added

- Ability to provide a callback to `setTime` from [@M4sterShake](https://github.com/M4sterShake).

### Fixed

- Invalid date toString() now returns "Invalid Date" from [@zoehneto](https://github.com/zoehneto).
- Setting 0 time works as other times.


timeshift.d.ts
export type TimeCallback = () => number | undefined;

/**
 * Set the current time in milliseconds after Jan 1 1970 in UTC time or a callback which
 * returns the time in this format. Setting this to undefined will reset to the real
 * current time.
 */
export function setTime(millis: number | TimeCallback | undefined): void;

/**
 * @return the currently overridden time value as milliseconds after Jan 1 1970 in UTC time.
 * The default value is undefined, which indicates using the real current time.
 */
export function getTime(): number | undefined;

/**
 * Set the time zone offset in minutes.  -60 corresponds to GMT+1, +60 to GTM-1.
 * Changing this will affect the results also for previously created Date instances.
 */
export function setTimezoneOffset(minutes: number): void;

/**
 * @return the current time zone offset in minutes.  A value of -60 corresponds to GMT+1,
 * +60 to GTM-1.  Default value is from new Date().getTimezoneOffset().
 */
export function getTimezoneOffset(): number;

/**
 * Access to the Mocked Date.
 */
export const Date: DateConstructor;

/**
 * Access to the original Date constructor.
 */
export const OriginalDate: DateConstructor;


timeshift.js
/*!
 * TimeShift.js
 *
 * Copyright 2013 Mobile Wellness Solutions MWS Ltd, Sampo Juustila
 * Released under the MIT license
 */

(function() {

	var root = this;
	var OriginalDate = root.Date;

	var TimeShift;
	if (typeof exports !== 'undefined') {
		TimeShift = exports;
	} else {
		TimeShift = root.TimeShift = {};
	}
	
	
	var currentTime = undefined;
	var timezoneOffset = new OriginalDate().getTimezoneOffset();
	
	function currentDate() {
		var time = TimeShift.getTime();
		if (typeof time === 'function') {
			time = time();
		}
		if (time !== undefined) {
			return new OriginalDate(time);
		} else {
			return new OriginalDate();
		}
	}
	

	function realLocalToUtc(realLocal) {
		return new OriginalDate(realLocal.getTime() - realLocal.getTimezoneOffset()*60*1000 + timezoneOffset*60*1000);
	}
	function utcToLocal(utc) {
		return new OriginalDate(utc.getTime() - timezoneOffset*60*1000);
	}
	function localToUtc(local) {
		return new OriginalDate(local.getTime() + timezoneOffset*60*1000);
	}
	function twoDigit(n) {
		if (n < 10) {
			return "0" + n;
		} else {
			return "" + n;
		}
	}
	function timezoneName() {
		var zone = "GMT";
		var offset = Math.abs(timezoneOffset);
		if (timezoneOffset < 0) {
			zone = zone + "+";
		} else if (timezoneOffset > 0) {
			zone = zone + "-";
		} else {
			return zone;
		}
		return zone + twoDigit(Math.floor(offset/60)) + twoDigit(offset%60);
	}

	
	/**
	 * Return the current time zone offset in minutes.  A value of -60 corresponds to GMT+1,
	 * +60 to GTM-1.  Default value is from new Date().getTimezoneOffset().
	 */
	TimeShift.getTimezoneOffset = function() {
		return timezoneOffset;
	}
	
	/**
	 * Set the time zone offset in minutes.  -60 corresponds to GMT+1, +60 to GTM-1.
	 * Changing this will affect the results also for previously created Date instances.
	 */
	TimeShift.setTimezoneOffset = function(offset) {
		timezoneOffset = offset;
	}
	
	/**
	 * Return the currently overridden time value set by `TimeShift.setTime`, or undefined if
	 * not overridden.
	 */
	TimeShift.getTime = function() {
		return currentTime;
	}

	/**
	 * Set the current time in milliseconds after Jan 1 1970 in UTC time, or a callback
	 * which returns the same.  Setting this to undefined will reset to the real current time.
	 */
	TimeShift.setTime = function(time) {
		currentTime = time;
	}
	
	/**
	 * Access to the original Date constructor.
	 */
	TimeShift.OriginalDate = OriginalDate;

	
	/**
	 * Mock implementation of Date.
	 */
	TimeShift.Date = function() {

		// Detect whether we're being called with 'new'
		// From http://stackoverflow.com/questions/367768/how-to-detect-if-a-function-is-called-as-constructor
		var isConstructor = false;
		if (this instanceof TimeShift.Date && !this.__previouslyConstructedByTimeShift) {
			isConstructor = true;
			this.__previouslyConstructedByTimeShift = true;
		}
		if (!isConstructor) {
			return (new TimeShift.Date()).toString();
		}

		switch (arguments.length) {
		case 0:
			this.utc = currentDate();
			break;
		case 1:
			this.utc = new OriginalDate(arguments[0]);
			break;
		case 2:
			this.utc = realLocalToUtc(new OriginalDate(arguments[0], arguments[1]));
			break;
		case 3:
			this.utc = realLocalToUtc(new OriginalDate(arguments[0], arguments[1], arguments[2]));
			break;
		case 4:
			this.utc = realLocalToUtc(new OriginalDate(arguments[0], arguments[1], arguments[2], arguments[3]));
			break;
		case 5:
			this.utc = realLocalToUtc(new OriginalDate(arguments[0], arguments[1], arguments[2], arguments[3], arguments[4]));
			break;
		case 6:
			this.utc = realLocalToUtc(new OriginalDate(arguments[0], arguments[1], arguments[2], arguments[3], arguments[4], arguments[5]));
			break;
		default:
			this.utc = realLocalToUtc(new OriginalDate(arguments[0], arguments[1], arguments[2], arguments[3], arguments[4], arguments[5], arguments[6]));
			break;
		}
	}
	
	TimeShift.Date.prototype.getDate = function() { return utcToLocal(this.utc).getUTCDate(); }
	TimeShift.Date.prototype.getDay = function() { return utcToLocal(this.utc).getUTCDay(); }
	TimeShift.Date.prototype.getFullYear = function() { return utcToLocal(this.utc).getUTCFullYear(); }
	TimeShift.Date.prototype.getHours = function() { return utcToLocal(this.utc).getUTCHours(); }
	TimeShift.Date.prototype.getMilliseconds = function() { return utcToLocal(this.utc).getUTCMilliseconds(); }
	TimeShift.Date.prototype.getMinutes = function() { return utcToLocal(this.utc).getUTCMinutes(); }
	TimeShift.Date.prototype.getMonth = function() { return utcToLocal(this.utc).getUTCMonth(); }
	TimeShift.Date.prototype.getSeconds = function() { return utcToLocal(this.utc).getUTCSeconds(); }
	
	TimeShift.Date.prototype.getUTCDate = function() { return this.utc.getUTCDate(); }
	TimeShift.Date.prototype.getUTCDay = function() { return this.utc.getUTCDay(); }
	TimeShift.Date.prototype.getUTCFullYear = function() { return this.utc.getUTCFullYear(); }
	TimeShift.Date.prototype.getUTCHours = function() { return this.utc.getUTCHours(); }
	TimeShift.Date.prototype.getUTCMilliseconds = function() { return this.utc.getUTCMilliseconds(); }
	TimeShift.Date.prototype.getUTCMinutes = function() { return this.utc.getUTCMinutes(); }
	TimeShift.Date.prototype.getUTCMonth = function() { return this.utc.getUTCMonth(); }
	TimeShift.Date.prototype.getUTCSeconds = function() { return this.utc.getUTCSeconds(); }
	
	TimeShift.Date.prototype.setDate = function() { var d = utcToLocal(this.utc); d.setUTCDate.apply(d, Array.prototype.slice.call(arguments, 0)); this.utc = localToUtc(d); return this.utc.getTime(); }
	TimeShift.Date.prototype.setFullYear = function() { var d = utcToLocal(this.utc); d.setUTCFullYear.apply(d, Array.prototype.slice.call(arguments, 0)); this.utc = localToUtc(d); return this.utc.getTime(); }
	TimeShift.Date.prototype.setHours = function() { var d = utcToLocal(this.utc); d.setUTCHours.apply(d, Array.prototype.slice.call(arguments, 0)); this.utc = localToUtc(d); return this.utc.getTime(); }
	TimeShift.Date.prototype.setMilliseconds = function() { var d = utcToLocal(this.utc); d.setUTCMilliseconds.apply(d, Array.prototype.slice.call(arguments, 0)); this.utc = localToUtc(d); return this.utc.getTime(); }
	TimeShift.Date.prototype.setMinutes = function() { var d = utcToLocal(this.utc); d.setUTCMinutes.apply(d, Array.prototype.slice.call(arguments, 0)); this.utc = localToUtc(d); return this.utc.getTime(); }
	TimeShift.Date.prototype.setMonth = function() { var d = utcToLocal(this.utc); d.setUTCMonth.apply(d, Array.prototype.slice.call(arguments, 0)); this.utc = localToUtc(d); return this.utc.getTime(); }
	TimeShift.Date.prototype.setSeconds = function() { var d = utcToLocal(this.utc); d.setUTCSeconds.apply(d, Array.prototype.slice.call(arguments, 0)); this.utc = localToUtc(d); return this.utc.getTime(); }
	
	TimeShift.Date.prototype.setUTCDate = function() { this.utc.setUTCDate.apply(this.utc, Array.prototype.slice.call(arguments, 0)); return this.utc.getTime(); }
	TimeShift.Date.prototype.setUTCFullYear = function() { this.utc.setUTCFullYear.apply(this.utc, Array.prototype.slice.call(arguments, 0)); return this.utc.getTime(); }
	TimeShift.Date.prototype.setUTCHours = function() { this.utc.setUTCHours.apply(this.utc, Array.prototype.slice.call(arguments, 0)); return this.utc.getTime(); }
	TimeShift.Date.prototype.setUTCMilliseconds = function() { this.utc.setUTCMilliseconds.apply(this.utc, Array.prototype.slice.call(arguments, 0)); return this.utc.getTime(); }
	TimeShift.Date.prototype.setUTCMinutes = function() { this.utc.setUTCMinutes.apply(this.utc, Array.prototype.slice.call(arguments, 0)); return this.utc.getTime(); }
	TimeShift.Date.prototype.setUTCMonth = function() { this.utc.setUTCMonth.apply(this.utc, Array.prototype.slice.call(arguments, 0)); return this.utc.getTime(); }
	TimeShift.Date.prototype.setUTCSeconds = function() { this.utc.setUTCSeconds.apply(this.utc, Array.prototype.slice.call(arguments, 0)); return this.utc.getTime(); }
	

	TimeShift.Date.prototype.getYear = function() { return this.getFullYear() - 1900; }
	TimeShift.Date.prototype.setYear = function(v) { return this.setFullYear(v + 1900); }
	
	TimeShift.Date.prototype.getTime = function() { return this.utc.getTime(); }
	TimeShift.Date.prototype.setTime = function(v) { return this.utc.setTime(v); }
	
	TimeShift.Date.prototype.getTimezoneOffset = function() { return timezoneOffset; }
	
	TimeShift.Date.prototype.toDateString = function() { return utcToLocal(this.utc).toDateString(); }  // Wrong
	TimeShift.Date.prototype.toLocaleDateString = function() { return utcToLocal(this.utc).toLocaleDateString(); }  // Wrong
	
	TimeShift.Date.prototype.toISOString = function() { return this.utc.toISOString(); }
	TimeShift.Date.prototype.toGMTString = function() { return this.utc.toGMTString(); }
	TimeShift.Date.prototype.toUTCString = function() { return this.utc.toUTCString(); }

	TimeShift.Date.prototype.toString = function() {
		var wkdays = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
		var months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
		var d = utcToLocal(this.utc);

		// Match browser behavior for invalid dates
		if (d.toString() === "Invalid Date") {
			return d.toString();
		}

		// Mon Mar 05 2012 06:07:08 GMT+0500
		return wkdays[d.getUTCDay()] + " " + months[d.getUTCMonth()] + " " + twoDigit(d.getUTCDate()) + " " + d.getUTCFullYear() +
			" " + twoDigit(d.getUTCHours()) + ":" + twoDigit(d.getUTCMinutes()) + ":" + twoDigit(d.getUTCSeconds()) + " " + timezoneName();
	}
	TimeShift.Date.prototype.toLocaleString = function() { return this.toString(); }  // Wrong
	TimeShift.Date.prototype.toLocaleTimeString = function() { return this.toString(); }  // Wrong
	TimeShift.Date.prototype.toTimeString = function() { return this.toString(); }  // Wrong

	TimeShift.Date.prototype.toJSON = function() { return this.utc.toJSON(); }
	TimeShift.Date.prototype.valueOf = function() { return this.utc.getTime(); }


	TimeShift.Date.now = function() { return currentDate().getTime(); }
	TimeShift.Date.parse = OriginalDate.parse;  // Wrong
	TimeShift.Date.UTC = OriginalDate.UTC;

	
	/**
	 * Helper method that describes a Date object contents.
	 */
	TimeShift.Date.prototype.desc = function() {
		return "utc=" + this.utc.toUTCString() + "   local=" + utcToLocal(this.utc).toUTCString() + "   offset=" + timezoneOffset;
	}

}).call(typeof global !== 'undefined' ? global : this);
